src/coreclr/jit/hwintrinsic.cpp

```diff
@@ -1148,10 +1148,72 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
             case 2:
                 op2 = getArgForHWIntrinsic(sigReader.GetOp2Type(), sigReader.op2ClsHnd);
                 op2 = addRangeCheckIfNeeded(intrinsic, op2, mustExpand, immLowerBound, immUpperBound);
-                op1 = getArgForHWIntrinsic(sigReader.GetOp1Type(), sigReader.op1ClsHnd);
 
-                retNode = isScalar ? gtNewScalarHWIntrinsicNode(retType, op1, op2, intrinsic)
-                                   : gtNewSimdHWIntrinsicNode(retType, op1, op2, intrinsic, simdBaseJitType, simdSize);
+#ifdef TARGET_ARM64
+                if ((intrinsic == NI_AdvSimd_VectorTableLookup) || (intrinsic == NI_AdvSimd_Arm64_VectorTableLookup))
+                {
+                    op1 = impPopStack().val;
+
+                    if (op1->TypeGet() == TYP_STRUCT)
+                    {
+                        assert(op1->OperIsLocal());
+
+                        LclVarDsc* op1VarDsc  = lvaGetDesc(op1->AsLclVar());
+                        unsigned   lclNum     = lvaGetLclNum(op1VarDsc);
+                        unsigned   fieldCount = info.compCompHnd->getClassNumInstanceFields(sigReader.op1ClsHnd);
+
+                        GenTreeFieldList* fieldList = new (this, GT_FIELD_LIST) GenTreeFieldList();
+                        int               offset    = 0;
+                        for (unsigned fieldId = 0; fieldId < fieldCount; fieldId++)
+                        {
+                            GenTreeLclFld* fldNode = gtNewLclFldNode(lclNum, TYP_SIMD16, offset);
+                            fieldList->AddField(this, fldNode, offset, TYP_SIMD16);
+
+                            offset += op1VarDsc->lvSize() / fieldCount;
+                        }
+                        op1 = fieldList;
+
+                        switch (fieldCount)
+                        {
+                            case 1:
+                                // NI_AdvSimd_VectorTableLookup
+                                // NI_AdvSimd_Arm64_VectorTableLookup
+                                break;
+                            case 2:
+                                // NI_AdvSimd_VectorTableLookup_2
+                                // NI_AdvSimd_Arm64_VectorTableLookup_2
+                                intrinsic = (NamedIntrinsic)(intrinsic + 1);
```

Comment:
Why do it this indirect way through addition instead of `intrinsic = NI_AdvSimd_Arm64_VectorTableLookup_2`, etc?

Reply:
because we want to set appropriate intrinsic based on if it is `AdvSimd` or `AdvSimd.Arm64`.

```c++
case 2:
  if (intrinsic == NI_AdvSimd_VectorTableLookup)
  {
     intrinsic = NI_AdvSimd_VectorTableLookup_2;
  }
  else
  {
     intrinsic = NI_AdvSimd_Arm64_VectorTableLookup_2;
  }
  break;
case 3:
  if (intrinsic == NI_AdvSimd_VectorTableLookup)
  {
     intrinsic = NI_AdvSimd_VectorTableLookup_3;
  }
  else
  {
     intrinsic = NI_AdvSimd_Arm64_VectorTableLookup_3;
  }
  break;
```